stout/composed-loop.h

```diff
@@ -0,0 +1,154 @@
+    }
+
+    void Ended() {
+      stored_k_->Done();
+    }
+
+    PrevK_* stored_k_;
+  };
+
+  template <typename ComposedK>
+  struct Composable {
+    template <typename Arg>
+    using ValueFrom = Arg;
+
+    template <typename Arg>
+    auto k() && {
+      return Continuation<ComposedK, Arg>{std::move(composed_k)};
+    }
+
+    ComposedK* composed_k;
+  };
+};
+
+template <typename K>
+auto ComposedLoopRouter(K* k) {
+  return _ComposedLoopRouter::Composable<K>{k};
+}
+
+////////////////////////////////////////////////////////////////////////
+
+struct _ComposedLoop {
+  template <typename K_, typename F_, typename Arg_>
+  struct Continuation : public TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, F_ f)
+      : k_(std::move(k)),
+        f_(std::move(f)) {}
+
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      // Executes when Body is called from the above-level stream.
+      if (innerStream_ == nullptr) {
+        inner_.emplace(Build(operator|(f_(std::forward<Args>(args)...), ComposedLoopRouter(this))));
+        innerStream_ = &inner_.value();
+        inner_.value().Start();
+      } else {
+        // Executes when Body is called from inner stream Next function.
+        k_.Body(std::forward<Args>(args)...);
+      }
+    }
+
+    void Ended() {
+      // When the topmost loop have no more values there no inner stream
+      // initialization, so it can indicate the loop end.
+      if (innerStream_ == nullptr) {
+        k_.Ended();
+        return;
+      }
+
+      // Trying to get new value from the loop above.
+      innerStream_ = nullptr;
+      stream_->Next();
+    }
+
+    void Next() override {
+      if (innerStream_ == nullptr) {
+        stream_->Next();
+        return;
+      }
+      innerStream_->Next();
+    }
+
+    void Done() override {
+      Ended();
+    }
+
+    using Left_ = typename std::invoke_result<F_, int>::type;
+    using Right_ = _ComposedLoopRouter::Composable<Continuation>;
+    using EArg_ = typename std::invoke_result<decltype(&operator|<Left_, Right_>), Left_, Right_>::type;
+    using E_ = typename std::invoke_result<decltype(&Build<EArg_>), EArg_>::type;
+
+    K_ k_;
+    F_ f_;
+
+    std::optional<E_> inner_;
+    TypeErasedStream* innerStream_ = nullptr;
+    TypeErasedStream* stream_ = nullptr;
+  };
+
+  template <typename F>
+  struct Composable {
+    template <typename Arg>
+    using ValueFrom = Arg;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, F, Arg>{std::move(k), std::move(f)};
+    }
+
+    F f;
+  };
+};
+
+template <typename F>
+auto ComposedLoop(F f) {
```

Comment:
I tried to put myself in a "first time seeing this" brain-mode. I found `ComposedLoop` to not be immediately obvious as a name, given what this thing does. There's a stream within a stream. What does that have to do with "loop"? And isn't everything in eventuals "composed"? 

How would you feel about renaming to... `InnerStream`? `NewStream`? `ForEach`?

Reply:
I like InnerStream?
What do you think, @benh ?

Reply:
Hmm, how about `StreamForEach`?

It reads pretty nice I think:

```
Iterate(std::vector<int>{1, 2, 3})
    | StreamForEach([](int i) {
      return Range(0, i);
    })
```

Reply:
`StreamForEach` looks good to me!